### PR TITLE
Workaround for duplicate task creation

### LINKED ISSUE #98: Duplicate tasks when specifying multiple func_params
When create a task with multiple `func_params` and connecting that task to any other task (whether it has parallelism in it or not) will double the run count of the original parallel task.

A simple example:

```
from hera.task import Task
from hera.workflow import Workflow
from hera.workflow_service import WorkflowService

from spincycle.runner.utils import get_sa_token


def use_secret(a):
    print(a)


# todo: replace the domain and token with your own
namespace = "etl-argo-v3"
token = get_sa_token(
    "default", namespace=namespace, config_file="/home/micha/.kube/config"
)

ws = WorkflowService(
    host="127.0.0.1:2746",
    token=token,
    namespace=namespace,
)
w = Workflow("volume-provision", ws)

t1 = Task(
    "use_secret1",
    use_secret,
    func_params=[{"a": "a"}, {"a": "a2"}],
)
w.add_task(t1)

t2 = Task(
    "use_secret2",
    use_secret,
    func_params=[{"a": "a"}],
)
t1 >> t2
w.add_task(t2)

ws.submit(w.workflow, namespace=namespace)
```

If you run and submit that, things will look fine within the argo UI:

image

However, if you look at the logs you see that in fact there were 4 pods run for `use-secret1`, two for each parameter choices:

```
Name:                volume-provision-3ab82d9e
Namespace:           etl
ServiceAccount:      default
Status:              Succeeded
Conditions:          
 PodRunning          False
 Completed           True
Created:             Wed Feb 16 18:19:19 +0100 (3 minutes ago)
Started:             Wed Feb 16 18:19:19 +0100 (3 minutes ago)
Finished:            Wed Feb 16 18:19:39 +0100 (3 minutes ago)
Duration:            20 seconds
Progress:            5/5
ResourcesDuration:   8s*(1 cpu),1m26s*(100Mi memory)

STEP                          TEMPLATE                   PODNAME                               DURATION  MESSAGE
 ✔ volume-provision-3ab82d9e  volume-provision-3ab82d9e                                                    
 ├─✔ use-secret1(0:a:"a")     use-secret1                volume-provision-3ab82d9e-940572814   4s          
 ├─✔ use-secret1(0:a:\"a\")   use-secret1                volume-provision-3ab82d9e-4054141412  5s          
 ├─✔ use-secret1(1:a:\"a2\")  use-secret1                volume-provision-3ab82d9e-4279694763  4s          
 ├─✔ use-secret1(1:a:"a2")    use-secret1                volume-provision-3ab82d9e-3966846017  4s          
 └─✔ use-secret2              use-secret2                volume-provision-3ab82d9e-1375103652  4s          
```

I've been trying to debug why this is happening but I can't seem to get anywhere. Taking out the `use-secret2` task or removing the parallelism for `use-secret1` leads to the expected number of tasks being run. A notable thing I've found is that the parallel task is always doubled in the number of tasks, no matter how parallel it's dependencies are. Also, the difference in quote-escaping between the original and duplicate tasks is consistent across all the times I've seen this happening... I presume it's a clue but I'm unsure where to trace it back to.

For reference, the template that is created is:

```
- name: volume-provision-3ab82d9e
  inputs: {}
  outputs: {}
  metadata: {}
  dag:
    tasks:
      - name: use-secret1
        template: use-secret1
        arguments:
          parameters:
            - name: a
              value: '{{item.a}}'
        withItems:
          - a: '"a"'
          - a: '"a2"'
      - name: use-secret2
        template: use-secret2
        arguments:
          parameters:
            - name: a
              value: '"a"'
        dependencies:
          - use-secret1
  parallelism: 50
- name: use-secret1
  inputs:
    parameters:
      - name: a
        value: '{{item.a}}'
  outputs: {}
  metadata: {}
  daemon: false
  script:
    name: use-secret1
    image: 'python:3.7'
    command:
      - python
    resources:
      limits:
        cpu: '1'
        memory: 4Gi
      requests:
        cpu: '1'
        memory: 4Gi
    source: |
      import json
      a = json.loads('{{inputs.parameters.a}}')

      print(a)
- name: use-secret2
  inputs:
    parameters:
      - name: a
        value: '"a"'
  outputs: {}
  metadata: {}
  daemon: false
  script:
    name: use-secret2
    image: 'python:3.7'
    command:
      - python
    resources:
      limits:
        cpu: '1'
        memory: 4Gi
      requests:
        cpu: '1'
        memory: 4Gi
    source: |
      import json
      a = json.loads('{{inputs.parameters.a}}')

      print(a)
```

### PR DESCRIPTION
**Pull request by mynameisfiber**
Due to argoproj/argo-workflows#7895, when creating parallel tasks using `withItems` where an item is a single-quote-surrounded json string (ie: `'"foo"'`), duplicate tasks get executed.

This PR is a workaround by avoiding sending json-encoded strings and instead using the native-string type supported by argo. This also introduces the need for error checking in the resulting script and may break backwards compatibility for anyone rewriting `Task.get_param_script_portion`.

Resolves #98 

Also, :confetti_ball: for 100th issue/pr!

### CODE REVIEW

File: src/hera/json_utils.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        return value
+    else:
+        return json.dumps(value)
+
+
+def create_param_extraction(param_name: str) -> str:
+    return f"""
+try:
+    {param_name} = json.loads('''{{{{inputs.parameters.{param_name}}}}}''')
+except json.JSONDecodeError:
```
Comment (flaviuvadan) [inline src/hera/json_utils.py:23]: Under what circumstances will `json.loads` throw a decode error here? Is the exception block dedicated to catching strings that are formatted as `'"s"'` vs '"'s'"', perhaps?

File: tests/test_task.py [hunk 2]
```diff
@@ -88,44 +89,37 @@ def test_param_getter_parses_single_param_val_on_base_model_payload(mock_model,
 def test_param_script_portion_adds_formatted_json_calls(op):
     t = Task('t', op, [{'a': 1}])
     script = t.get_param_script_portion()
-    assert script == 'import json\na = json.loads(\'{{inputs.parameters.a}}\')\n'
+    assert script == f"import json\n{create_param_extraction('a')}"
```
Comment (flaviuvadan) [inline tests/test_task.py:92]: I love the replacements! 🚀 What do you think about including some small tests for `create_param_extraction` independently?

**Comment by flaviuvadan** [comment]
> Also, 🎊 for 100th issue/pr!

This is very kind of you to notice and bring up 😃

File: src/hera/json_utils.py (hunk 1 above)
Comment (mynameisfiber) [reply src/hera/json_utils.py:23]: Strings will always come through without any enclosing quotes, leading to the decode error. This is a result of [this line](https://github.com/argoproj-labs/hera-workflows/pull/100/files#diff-c5438bb6fa034bef3ea5a671e3609f3b0ce4caabe5656501d546ca2537d133bcR13-R14)

File: tests/test_task.py (hunk 2 above)
Comment (mynameisfiber) [reply tests/test_task.py:92]: woo! i could definitely do that. is there an edge case the current tests in `test_task.py` don't cover?

File: src/hera/task.py [hunk 3]
```diff
@@ -614,10 +609,11 @@ def get_parallel_items(self) -> List[_Item]:
...
             for k, v in func_param.items():
-                if isinstance(v, BaseModel):
-                    item[k] = v.json()
-                else:
-                    item[k] = json.dumps(v)  # type: ignore
+                v_json = encode_json(v)
+                try:
+                    item[k] = json.loads(v_json)
+                except json.JSONDecodeError:
+                    item[k] = v_json
```
[the remark this answers is not in the document]
Comment (mynameisfiber) [reply src/hera/task.py:616]: Yea! I was wondering the same thing about using `BaseModel.dict()` as well. I was unsure about how pydantic would behave if we had a model that contains other models in it. I did the loads/dumps routine in order to ensure the resulting object was 100% native json-compatible types. If `.dict()` can make the same guarantee then I definitely think you're way is much nicer!

In terms of the argo side question, I'm not 100% sure. In the argo spec, `withItems` expects a list of maps, so with a single item being submitted I assume argo would see see a list of maps, just the list only contains a single item.

[the remark this answers is not in the document]
Comment (flaviuvadan) [reply src/hera/task.py:616]: The `withItems` field is very finicky, actually... which is why I was asking about the Argo side. The `_Item` was introduced to overcome an open problem in OpenAPI SDK generation. Any chance you have a cluster available to test these changes in? Specifically, the hello world and parallel diamonds example - those are the two core ones I tested when `_Item` was introduced to makes sure things are serialized properly. Alternatively, I can help with this as well as testing can be a bit time consuming since it's best to assemble some tasks with a variety of parameters, such as a string, an integer, and a Pydantic model to test serialization in the single task and parallel task case.

[the remark this answers is not in the document]
Comment (mynameisfiber) [reply src/hera/task.py:616]: I will test with those workflows tomorrow morning!

I will say however that will this change I have been able to submit workloads to my cluster that have very complex `withItems` (strings, ints and dicts) and combinations of multiple and single `func_params`.

[the remark this answers is not in the document]
Comment (flaviuvadan) [reply src/hera/task.py:616]: Oh, that sounds amazing! Thank you for already testing that 🙂 I should also say thank you for using Hera ❤️ Very excited to hear that you are using it to submit fairly complex workflows! If you have the chance then, I guess testing the single pod/single task workflow would be sufficient. At @dynotx we had some serialization issues when submitting `pandas.DataFrame` queries through configs and we had to futz around with the strings that are submitted, which caused problems/differences between the way we submitted single vs. parallel tasks, but we reverted to the current implementation and used it differently 🤷🏻 want to ensure the same thing does not happen here.

**Comment by flaviuvadan** [comment]
@mynameisfiber any updates on this? I am wondering if this is still an issue given the discussion on [the issue](https://github.com/argoproj/argo-workflows/issues/7895)?

**Comment by mynameisfiber** [comment]
@flaviuvadan sorry for the delay! things got stalled on this project but I'll be back on it next week.

**Comment by mynameisfiber** [comment]
@flaviuvadan hello! so indeed this whole issues stems from multiple instances without `--managed-namespace` having undefined behaviour. There [does seem to be](https://github.com/argoproj/argo-workflows/issues/7801) efforts in trying to fix this, so I leave it up to you as to whether it is worth having a fix on hera's side in place. Alternatively, we could detect the potential for this edge-case and issue a `warnings.warn` to help users not get caught in the same impossible bug-tracking I was in.

**Comment by flaviuvadan** [comment]
@mynameisfiber a recent contributor added the triple quotes to get around some deserialization errors, which strikes me as a medium between the solution illustrated here and what needs to be fixed on Argo's server side. Let's close this for now and see how that issue evolves. I will keep an eye out and if it gets resolved, or re-raised again on Hera's side, we can go from there. Thank you so much for the investigation! 🙂  Hope to see more contributions from you in the future,
